source/MaterialXCore/Document.cpp

```diff
@@ -357,6 +357,14 @@ vector<OutputPtr> Document::getMaterialOutputs() const
 
 vector<NodeDefPtr> Document::getMatchingNodeDefs(const string& nodeName) const
 {
+    // Return all nodedefs from datalibrary if available
```

Comment:
@ashwinbhat Wouldn't we also need to handle the case of a content document that contains _its own_ custom nodes, as might be packaged and shared between studios?  To handle that case, we'd want to first check for a matching nodedef in the data libraries, and then fall back to checking for a matching nodedef in the content document.

Following that pattern, perhaps we should take the same approach with the other updated methods in this pull request, including `Document::getMatchingImplementation` below.

Reply:
@jstone-lucasfilm in the code below, we check in the datalibrary if available and then fallback to checking in local.

Reply:
I see what you mean, but couldn't there be nodedefs for the given node name in _both_ the data library and the content document?  For example, the data library might have a node named `multiply`, but only the content document would have the specific version  of `multiply` that the caller is looking for.

Reply:
The proposed workflow, if the caller chooses to Register a datalibrary then the data library gets precedence. The caller can still continue to use the import mechanism. 

We can update getMatchingNodeDefs so that it checks both data library and content document. Would that be preferred?

Reply:
@ashwinbhat I'm imagining that we want the behavior to exactly match what a user would see if they called `importLibrary` instead of `registerDataLibrary`, and I believe that would require us to check both the data library and content document, merging the matches from both of them.

Let me know your thoughts, though, and that's just one opinion on the matter!

Reply:
I think this is an interesting point to discuss - should `registerDataLibrary` match the `importLibrary` behavior?  

I can certainly imagine circumstances where I might even want to not fallback to any local node definition in the local document and want to be able to validate that my document only uses nodes from the dataLibrary provided. 

Perhaps `registerDataLibrary()` has some concept of strictness?  or some input that could control its "fallback" behavior?

Reply:
I've updated this so that both the data library and content document are now checked. @ld-kerley I think once the datalibrary concept is adopted we could start introducing strictness. The current goal is keep it similar to importLibrary workflow.